[TIMOB-20320] Android: Provide generic requestPermission(s) method

stgrosshh opened the pull request:
Add generic permissions request handling for Android Marshmallow
see example usage in tipermissions module
https://github.com/stgrosshh/tipermissions

JIRA: https://jira.appcelerator.org/browse/TIMOB-20320

stgrosshh committed:
Proposal for TIMOB-20320
Add generic permissions request handling for Android Marshmallow
see example usage in tipermissions module
https://github.com/stgrosshh/tipermissions

AppWerft commented:
It is very handy to use. Actually we only need requestPermissions(). hasPermission() is only for static test.

Here my test: 

`require('ti.permissions').requestPermissions(['android.permission.RECORD_AUDIO', 'android.permission.WRITE_EXTERNAL_STORAGE'], 0, function(e) {
	if (e.success == !0) startAudioRecorder();
	else alert(L('USER_HAS_DENIED'));
});`

stgrosshh committed:
Fix possible NPE issue
Change to strong references to avoid some issues with async UI creation
Provide infos about requested and denied permissions in callback result

ashcoding commented on android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java:
```diff
@@ -475,6 +492,82 @@ public void onRequestPermissionsResult(int requestCode,
+     */
+	public void registerPermissionRequestCallback(Integer requestCode,
+														 KrollFunction callback,KrollObject context,
+														 String... requestedPermissions) {
```
Line 558 to Line 563, Slight formatting issue.

ashcoding commented on android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java:
```diff
@@ -475,6 +492,82 @@ public void onRequestPermissionsResult(int requestCode,
 		}
 	}
 
+	private void handlePermissionRequestResult(Integer requestCode,String[] permissions, int[] grantResults) {
```
Formatting issue for the params.

ashcoding commented:
There seems to be slight formatting issues here and there in the code. Could you format the code?
Line 558 to 564. Line 460. Line 495. Line 112. Line 121.
You can refer to this to fix it up http://docs.appcelerator.com/platform/latest/#!/guide/Java_Coding_Standards

Or if you're using an IDE with some format function, that might help speed things up. :)

stgrosshh committed:
Code formatting as requested

stgrosshh commented:
Done formatting as requested.

hansemannn commented:
@stgrosshh Can you add the docs for your new method? Also, how does this affect current apps? We have to ensure backwards-compatibility by deprecating the old API. Thank you!

stgrosshh commented:
@hansemannn for documentation pls see the provided JavaDoc. It's meant as a means for module developers primarely. See my tipermissions module https://github.com/stgrosshh/tipermissions for an example of usage.

I deprecated the old permission specific methods in my PR so there would be time for migration. I think some of the Appcelerator provided modules use those right now. Therefore some migration effort would be required though. In the meantime they still work of course.

hieupham007 commented:
Thanks for your contributions! I've refactored the code so we can remove the deprecated internal methods. Please refer to this PR: https://github.com/appcelerator/titanium_mobile/pull/7948
